File: lib/data/firestore.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

/// Firestore over its REST API, read without credentials: the security
/// rules let anyone read what the website shows (published posts), so the
/// app needs no Firestore SDK or session for them.
///
/// `runQuery` answers in Firestore's typed JSON (`{"stringValue": ...}`,
/// `{"mapValue": {"fields": ...}}`); [decodeValue] turns that into plain
/// Dart values.
class FirestoreClient {
  FirestoreClient({required this.projectId, http.Client? client})
    : _client = client ?? http.Client();

  final String projectId;
  final http.Client _client;

  Uri get runQueryUri => Uri.https(
    'firestore.googleapis.com',
    '/v1/projects/$projectId/databases/(default)/documents:runQuery',
  );

  /// Runs a structured query and returns the matching documents' fields,
  /// each with its `id` (the last part of the document path) added.
  Future<List<Map<String, dynamic>>> runQuery(
    Map<String, Object?> structuredQuery,
  ) async {
    final response = await _client.post(
      runQueryUri,
      headers: const {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
      body: jsonEncode({'structuredQuery': structuredQuery}),
    );
    if (response.statusCode != 200) {
      throw FirestoreException(
        'Firestore returned HTTP ${response.statusCode}',
      );
    }
    final rows = jsonDecode(response.body);
    if (rows is! List) throw FirestoreException('Unexpected Firestore reply');
    final docs = <Map<String, dynamic>>[];
    for (final row in rows.whereType<Map>()) {
      final error = row['error'];
      if (error is Map) {
        throw FirestoreException(error['message'] as String? ?? 'query failed');
      }
      final document = row['document'];
      if (document is! Map) continue;
      final fields = document['fields'];
      final name = document['name'] as String? ?? '';
      docs.add({
        ...decodeFields(fields is Map ? fields : const {}),
        'id': name.split('/').last,
      });
    }
    return docs;
  }
}

class FirestoreException implements Exception {
  FirestoreException(this.message);

  final String message;

  @override
  String toString() => message;
}

/// A Firestore value as the plain value it stands for: strings, numbers,
/// booleans, null, lists and maps; timestamps stay ISO 8601 strings.
Object? decodeValue(Object? value) {
  if (value is! Map) return null;
  if (value.containsKey('nullValue')) return null;
  if (value.containsKey('stringValue')) return value['stringValue'];
  if (value.containsKey('booleanValue')) return value['booleanValue'];
  if (value.containsKey('integerValue')) {
    return int.tryParse(value['integerValue'].toString());
  }
  if (value.containsKey('doubleValue')) {
    return (value['doubleValue'] as num?)?.toDouble();
  }
  if (value.containsKey('timestampValue')) return value['timestampValue'];
  if (value.containsKey('arrayValue')) {
    final values = (value['arrayValue'] as Map?)?['values'];
    return values is List ? values.map(decodeValue).toList() : <Object?>[];
  }
  if (value.containsKey('mapValue')) {
    final fields = (value['mapValue'] as Map?)?['fields'];
    return decodeFields(fields is Map ? fields : const {});
  }
  return null;
}

Map<String, dynamic> decodeFields(Map<dynamic, dynamic> fields) => {
  for (final entry in fields.entries)
    entry.key.toString(): decodeValue(entry.value),
};

/// Firestore's typed JSON for a plain value, for query filters.
Map<String, Object?> encodeValue(Object? value) => switch (value) {
  null => {'nullValue': null},
  bool b => {'booleanValue': b},
  int i => {'integerValue': '$i'},
  double d => {'doubleValue': d},
  String s => {'stringValue': s},
  List l => {
    'arrayValue': {'values': l.map(encodeValue).toList()},
  },
  _ => throw ArgumentError('cannot encode ${value.runtimeType} for Firestore'),
};

File: site/src/lib/firestore.ts
// Firestore over its REST API, read at build time without credentials:
// the security rules let anyone read what the website shows (published
// posts). No SDK, no service account; the project is the only setting.
//
// The response of `runQuery` is Firestore's typed JSON (`{stringValue}`,
// `{mapValue: {fields}}`, ...); `decode` turns it into plain objects.

/** The Firebase project the site is built from; set in astro.config.mjs. */
export const FIREBASE_PROJECT: string = import.meta.env.PUBLIC_FIREBASE_PROJECT;

const DATABASE = `projects/${FIREBASE_PROJECT}/databases/(default)`;
const ENDPOINT = `https://firestore.googleapis.com/v1/${DATABASE}/documents`;

type Value =
  | { nullValue: null }
  | { booleanValue: boolean }
  | { integerValue: string }
  | { doubleValue: number }
  | { stringValue: string }
  | { timestampValue: string }
  | { arrayValue: { values?: Value[] } }
  | { mapValue: { fields?: Record<string, Value> } };

/** A Firestore value as the plain JavaScript value it stands for. */
export function decode(value: Value): unknown {
  if ('nullValue' in value) return null;
  if ('booleanValue' in value) return value.booleanValue;
  if ('integerValue' in value) return Number(value.integerValue);
  if ('doubleValue' in value) return value.doubleValue;
  if ('stringValue' in value) return value.stringValue;
  if ('timestampValue' in value) return value.timestampValue;
  if ('arrayValue' in value) return (value.arrayValue.values ?? []).map(decode);
  if ('mapValue' in value) return decodeFields(value.mapValue.fields ?? {});
  throw new Error(`firestore: unknown value ${JSON.stringify(value)}`);
}

function decodeFields(fields: Record<string, Value>): Record<string, unknown> {
  return Object.fromEntries(Object.entries(fields).map(([key, value]) => [key, decode(value)]));
}

export interface Filter {
  field: string;
  op: 'EQUAL' | 'NOT_EQUAL' | 'LESS_THAN' | 'GREATER_THAN' | 'ARRAY_CONTAINS';
  value: string | number | boolean;
}

/** A document's fields plus its id (the last part of its path). */
export type Document = Record<string, unknown> & { id: string };

const encodeValue = (value: Filter['value']): Value =>
  typeof value === 'string'
    ? { stringValue: value }
    : typeof value === 'boolean'
      ? { booleanValue: value }
      : Number.isInteger(value)
        ? { integerValue: String(value) }
        : { doubleValue: value };

/**
 * Every document of a top-level collection matching all of `filters`.
 * Equality filters alone need no composite index, so callers sort the
 * result themselves rather than asking Firestore to.
 */
export async function query(collection: string, filters: Filter[]): Promise<Document[]> {
  if (!FIREBASE_PROJECT) throw new Error('firestore: PUBLIC_FIREBASE_PROJECT is not set');
  const where =
    filters.length === 1
      ? fieldFilter(filters[0])
      : { compositeFilter: { op: 'AND', filters: filters.map(fieldFilter) } };
  const response = await fetch(`${ENDPOINT}:runQuery`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ structuredQuery: { from: [{ collectionId: collection }], where } }),
  });
  const body = await response.text();
  if (!response.ok) throw new Error(`firestore: ${collection} query failed with HTTP ${response.status}: ${body}`);
  const rows: { document?: { name: string; fields?: Record<string, Value> }; error?: { message: string } }[] =
    JSON.parse(body);
  const failed = rows.find((row) => row.error);
  if (failed?.error) throw new Error(`firestore: ${collection} query failed: ${failed.error.message}`);
  return rows
    .filter((row): row is { document: { name: string; fields?: Record<string, Value> } } => !!row.document)
    .map(({ document }) => ({ ...decodeFields(document.fields ?? {}), id: document.name.split('/').pop() ?? '' }));
}

const fieldFilter = ({ field, op, value }: Filter) => ({
  fieldFilter: { field: { fieldPath: field }, op, value: encodeValue(value) },
});
